Support new upsample in symbolic, caffe2 backend & caffe2 frontend

zrphercule opened the pull request:
We updated the description of upsample_op in onnx: https://github.com/onnx/onnx/pull/1467
Therefore, we need to support the new upsample_op in caffe2-onnx backend as well.

houseroad commented on caffe2/onnx/helper.h:
```diff
@@ -12,6 +12,7 @@ namespace onnx {
 
 using ::ONNX_NAMESPACE::AttributeProto;
 using ::ONNX_NAMESPACE::NodeProto;
+using ::ONNX_NAMESPACE::TensorProto;
```
Using TensorProto here is not recommended, especially, it's a header file.
And another reason is that we have both caffe2::TensorProto and onnx::TensorProto (caffe2 does not have NodeProto and AttributeProto). I would suggest explicitly use onnx::TensorProto in the following code.

zrphercule replied:
Yeah that's a good idea.

zrphercule replied:
wait, you mean ONNX_NAMESPACE::TensorProto bot onnx::TensorProto, right?

houseroad commented on caffe2/onnx/helper.h:
```diff
@@ -71,6 +72,25 @@ inline AttributeProto MakeAttribute(
+  AttributeProto attr;
+  attr.set_name(name);
+  std::string val_buf;
+  *(attr.mutable_t()) = val;
```
Use CopyFrom instead of direct assign

houseroad commented on caffe2/onnx/helper.h:
```diff
@@ -71,6 +72,25 @@ inline AttributeProto MakeAttribute(
+inline AttributeProto MakeAttribute(const std::string& name, TensorProto& val) {
+  AttributeProto attr;
+  attr.set_name(name);
+  std::string val_buf;
```
unused?

houseroad commented on caffe2/onnx/helper.h:
```diff
@@ -71,6 +72,25 @@ inline AttributeProto MakeAttribute(
+  TensorProto ret;
+  ret.set_name(name);
+  ret.add_dims(v.size());
+  ret.set_data_type(TensorProto::FLOAT);
```
Can we make it general? vector template with char, and pass the data_type as parameter?

zrphercule replied:
I guess we can, if data_type is passed as parameter.

houseroad reviewed (commented):
Some inline comments

zrphercule commented:
@houseroad Any more comments? CI is green.

houseroad commented:
import and update the submodule internally?

zrphercule committed:
Support new upsample in symbolic, caffe2 backend & caffe2 frontend (#13272)

Summary:
We updated the description of upsample_op in onnx: https://github.com/onnx/onnx/pull/1467
Therefore, we need to support the new upsample_op in caffe2-onnx backend as well.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13272

Differential Revision: D12833656

fbshipit-source-id: d534b440b6ffbc1a8e91a573bcdefc31f3ef847f
